# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/WebGLContext.h
# Commit: b9f43d8a6c2d
# Full Hash: b9f43d8a6c2d9504dde533d3db13111ae4d31d06
# Author: Lee Salzman <lsalzman@mozilla.com>
# Date: 2022-07-12 21:56:41
# Regressor Bug: 1777426
# File Overlap Count: 2
# Description:
#   Bug 1777426 - Support fast readbacks in CopyToSwapChain with async present. r=sotaro,jgilbert
#   
#   Currently CopyToSwapChain creates spurious copies of the back buffer when SharedSurfaces aren't exportable (= ToSurfaceDescriptor returns Nothing from SharedSurface_Basic). These then later get read back into a CPU memory buffer when PresentFrontBufferToCompositor is used to send the buffer to RemoteTextureMap. This has associated performance and memory costs.
#   
#   Conceptually, we want Present/CopyToSwapChain to just do the right thing and automatically push buffers to RemoteTextureMap, rather than secondarily needing a hidden call to PresentFrontBufferToCompositor. Then we can get rid of the need to create front buffers whose only purpose is to shuttle data to PresentFrontBufferToCompositor which then shuttles it RemoteTextureMap.
# ==============================================================================

diff -r 0a598b1533e6 -r b9f43d8a6c2d dom/canvas/WebGLContext.h
--- a/dom/canvas/WebGLContext.h	Tue Jul 12 06:49:11 2022 +0000
+++ b/dom/canvas/WebGLContext.h	Tue Jul 12 06:56:19 2022 +0000
@@ -98,6 +98,7 @@
 
 namespace layers {
 class CompositableHost;
+class RemoteTextureOwnerClient;
 class SurfaceDescriptor;
 }  // namespace layers
 
@@ -491,7 +492,9 @@
   // the back buffer may be invalidated by this swap with the front buffer,
   // unless overriden by explicitly setting the preserveDrawingBuffer option,
   // which may incur a further copy to preserve the back buffer.
-  void Present(WebGLFramebuffer*, layers::TextureType, const bool webvr);
+  void Present(
+      WebGLFramebuffer*, layers::TextureType, const bool webvr,
+      const webgl::SwapChainOptions& options = webgl::SwapChainOptions());
   // CopyToSwapChain forces a copy from the supplied framebuffer into the back
   // buffer before swapping the front and back buffers of the swap chain for
   // compositing. The formats of the framebuffer and the swap chain buffers
@@ -514,6 +517,8 @@
   Maybe<uvec2> FrontBufferSnapshotInto(
       const std::shared_ptr<gl::SharedSurface>& front,
       const Maybe<Range<uint8_t>>);
+  Maybe<uvec2> SnapshotInto(GLuint srcFb, const gfx::IntSize& size,
+                            const Range<uint8_t>& dest);
   gl::SwapChain* GetSwapChain(WebGLFramebuffer*, const bool webvr);
   Maybe<layers::SurfaceDescriptor> GetFrontBuffer(WebGLFramebuffer*,
                                                   const bool webvr);
@@ -1233,6 +1238,12 @@
   gl::SwapChain mSwapChain;
   gl::SwapChain mWebVRSwapChain;
 
+  RefPtr<layers::RemoteTextureOwnerClient> mRemoteTextureOwner;
+
+  bool PushRemoteTexture(WebGLFramebuffer*, gl::SwapChain&,
+                         std::shared_ptr<gl::SharedSurface>,
+                         const webgl::SwapChainOptions& options);
+
   // --
 
   bool EnsureDefaultFB();